pigpen-cascading/src/main/clojure/pigpen/cascading.clj

```diff
@@ -41,335 +68,228 @@
+  (fn [{:keys [type]} ancestors flowdef] type))
+
+(s/defmethod command->flowdef :load
+  [{:keys [id fields], :as command} :- m/Load
+   _
+   ^FlowDef flowdef]
+  (let [^Tap tap (get-tap command)
+        pipe (->
+               (Pipe. (str id))
+               (Rename. (.getSourceFields tap) (cfields fields)))]
+    ;; side effect
+    (.addSource flowdef pipe tap)
+    pipe))
+
+(s/defmethod command->flowdef :store
+  [command :- m/Store
+   [{:keys [^Pipe pipe]}]
+   ^FlowDef flowdef]
+  (let [^Tap sink (get-tap command)]
+    ;; side effect
+    (.addTailSink flowdef pipe sink)
+    nil))
+
+(s/defmethod command->flowdef :reduce
+  [command :- m/Reduce
+   [{:keys [^Pipe pipe]}]
+   _]
+  (GroupBy. pipe Fields/NONE))
+
+(s/defmethod command->flowdef :reduce-fold
+  [{:keys [reduce :- m/Reduce
+           fold :- m/Mapcat]}
+   [{:keys [^Pipe pipe]}]
+   _]
+  (let [projections (:projections fold)
+        context (pr-str `'{:projections ~(prepare-projections projections)})
+        old-fields (cfields (get-in projections [0 :expr :args]))
+        new-fields (cfields (get-in projections [0 :alias]))]
+    (->
+      (PigPenAggregateBy. context pipe Fields/NONE old-fields)
+      (Rename. old-fields new-fields))))
+
+(s/defmethod command->flowdef :group
+  [{:keys [id keys fields join-types opts]} :- m/Group
+   ancestors
+   _]
+  (let [join-keys (group-key-cfields keys (:join-nils opts))
+        pipes (pipe-array
+                (if (:join-nils opts)
+                  (map :pipe ancestors)
+                  ;; This is to induce pigpen's nil-joining behavior, which
+                  ;; treats nils from the same relation as equal
+                  (->> ancestors
+                    (map-indexed
+                      (fn [i a]
+                        (let [fields (-> a :ancestor :fields cfields)]
+                          (Each. ^Pipe (:pipe a)
+                                 (InduceSentinelNils. (int i) fields))))))))]
+    (CoGroup. (str id) pipes join-keys Fields/NONE (BufferJoin.))))
+
+(s/defmethod command->flowdef :group-fold
+  [{:keys [group :- m/Group
+           fold :- m/Mapcat]}
+   ancestors
+   _]
+  (let [{:keys [id keys]} group
+        join-keys (group-key-cfields keys (get-in group [:opts :join-nils]))
+        ;; Perform all fold aggregations before co-grouping the relations
+        pipes (pipe-array
+                (zip [p (-> fold :projections next)
+                      {:keys [pipe]} ancestors
+                      key keys]
+                  (if (some-> p :expr :udf #{:fold})
+                    (let [context (pr-str `'{:projections ~(prepare-projections [p])})
+                          group-fields (cfields [key])
+                          arg-fields (cfields (get-in p [:expr :args]))]
+                      (PigPenAggregateBy. context pipe group-fields arg-fields))
+                    pipe)))]
+    (CoGroup. (str id) pipes join-keys Fields/NONE (BufferJoin.))))
```

Comment:
I don't think we want a CoGroup here. AggregateBy already does a GroupBy internally, so a CoGroup will add an extra map-reduce job that is not really necessary.